src/BSON/DBPointer.c

```diff
@@ -0,0 +1,386 @@
+		return false;
+	}
+
+	intern->namespace = estrndup(namespace, namespace_len);
+	intern->namespace_len = namespace_len;
+	
+	oid_intern = Z_OBJECTID_OBJ_P(oid);
+
+	strncpy(intern->oid, oid_intern->oid, sizeof(intern->oid));
+
+	return true;
+} /* }}} */
+
+/* Initialize the object from a HashTable and return whether it was successful.
+ * An exception will be thrown on error. */
+static bool php_phongo_dbpointer_init_from_hash(php_phongo_dbpointer_t *intern, HashTable *props TSRMLS_DC) /* {{{ */
+{
+#if PHP_VERSION_ID >= 70000
+	zval *namespace, *oid;
+
+	if ((namespace = zend_hash_str_find(props, "namespace", sizeof("namespace")-1)) && Z_TYPE_P(namespace) == IS_STRING &&
+		(oid = zend_hash_str_find(props, "oid", sizeof("oid")-1)) && Z_TYPE_P(oid) == IS_OBJECT && instanceof_function(Z_OBJCE_P(oid), php_phongo_objectid_interface_ce TSRMLS_CC)) {
+		return php_phongo_dbpointer_init(intern, Z_STRVAL_P(namespace), Z_STRLEN_P(namespace), oid TSRMLS_CC);
+	}
+#else
+	zval **namespace, **oid;
+
+	if (zend_hash_find(props, "namespace", sizeof("namespace"), (void**) &namespace) == SUCCESS && Z_TYPE_PP(namespace) == IS_STRING &&
+		zend_hash_find(props, "oid", sizeof("oid"), (void**) &oid) == SUCCESS && Z_TYPE_PP(oid) == IS_OBJECT && instanceof_function(Z_OBJCE_PP(oid), php_phongo_objectid_interface_ce TSRMLS_CC)) {
+
+		return php_phongo_dbpointer_init(intern, Z_STRVAL_PP(namespace), Z_STRLEN_PP(namespace), *oid TSRMLS_CC);
+	}
+#endif
+
+	phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "%s initialization requires \"namespace\" string field and \"oid\" MongoDB\\BSON\\ObjectID field", ZSTR_VAL(php_phongo_dbpointer_ce->name));
+	return false;
+} /* }}} */
+
+/* {{{ proto void MongoDB\BSON\DBPointer::__set_state(array $properties)
+*/
+static PHP_METHOD(DBPointer, __set_state)
+{
+	php_phongo_dbpointer_t *intern;
+	HashTable               *props;
+	zval                    *array;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "a", &array) == FAILURE) {
+		RETURN_FALSE;
+	}
+
+	object_init_ex(return_value, php_phongo_dbpointer_ce);
+
+	intern = Z_DBPOINTER_OBJ_P(return_value);
+	props = Z_ARRVAL_P(array);
+
+	php_phongo_dbpointer_init_from_hash(intern, props TSRMLS_CC);
+} /* }}} */
+
+/* {{{ proto string MongoDB\BSON\DBPointer::__toString()
+   Return the DBPointer's namespace string. */
+static PHP_METHOD(DBPointer, __toString)
+{
+	php_phongo_dbpointer_t *intern;
+	char                    *retval;
+	int                      retval_len;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		return;
+	}
+
+	intern = Z_DBPOINTER_OBJ_P(getThis());
+
+	retval_len = spprintf(&retval, 0, "[%s/%s]", intern->namespace, intern->oid);
+	PHONGO_RETVAL_STRINGL(retval, retval_len);
+	efree(retval);
+} /* }}} */
+
+/* {{{ proto string MongoDB\BSON\DBPointer::serialize()
+*/
+static PHP_METHOD(DBPointer, serialize)
+{
+	php_phongo_dbpointer_t  *intern;
```

Comment:
Indentation is also off by one space here. Is that an issue for all `php_phongo_dbpointer_t` variables?